### PR TITLE
MDEV-39466 proxy_protocol: fix off-by-one stack buffer overflow in PROXY v1 header parsing

### PR DESCRIPTION
**Pull request by uwezkhan**
Jira: MDEV-39466
The header parsing loop could append a null terminator past the end of
the buffer when processing maximum-length PROXY v1 headers.

Fix by increasing the buffer size by one byte and using bounded reads
while preserving full header parsing.

Treat zero-length reads as connection close in this context.

Add a regression test for maximum-length PROXY headers.

### CODE REVIEW

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a preliminary review.

As with the other one, please make sure a jira is present and describes the issue well.

Please:

* make sure the PR compiles and runs tests to completion
* please target the main branch: targeting development branches is not useful
* please revert the space only changes. 
* please consider making a regression test of some sort. Or explain why it cannot be done. If you need help please reach out to the developers on zulip.

**Comment by uwezkhan** [comment]
Thanks for the review.

I’ve added Jira issue MDEV-39466 and linked it to this PR.

I’ll update the remaining requested cleanup shortly.

File: mysql-test/main/proxy_protocol_v1_malformed.test [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+# This test sends crafted proxy protocol V1 headers over raw TCP
+# to verify that the server correctly rejects:
+#   A) Headers longer than MAX_PROXY_HEADER_LEN without a newline
+#   B) Fields longer than MAX_PROXY_HEADER_LEN
+#
+# The server must close the connection gracefully in all cases (no crash).
+#
+
+--source include/not_embedded.inc
+--source include/not_windows.inc
```
Comment (grooverdan) [inline mysql-test/main/proxy_protocol_v1_malformed.test:13]: why not Windows?

Can move `not_embedded.inc` to start of file.

File: mysql-test/main/proxy_protocol_v1_malformed.test [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+connect(SOCK, pack_sockaddr_in($ENV{MASTER_MYPORT}, inet_aton("127.0.0.1")))
+  or die "connect: $!\n";
+syswrite(SOCK, $hdr);
+select(undef, undef, undef, 0.5);
+close(SOCK);
+print "Test B: server did not crash after excessively long fields in V1 header\n";
+EOF
+
+# Allow time for the server to clean up aborted connections
+--sleep 1
```
Comment (grooverdan) [inline mysql-test/main/proxy_protocol_v1_malformed.test:66]: sleeps become fragile in CI systems. Use `--ping` to push and receive an OK, This will ensure the same connection has finished cleanup before this no-op command is processed.

Comment (grooverdan) [reply mysql-test/main/proxy_protocol_v1_malformed.test:66]: As a bonus, this verfies the server is still alive.

File: tests/mysql_client_test.c [hunk 3]
```diff
@@ -1488,7 +1488,7 @@ static void test_prepare()
   /* now, execute the prepared statement to insert 10 records.. */
   for (tiny_data= 0; tiny_data < 100; tiny_data++)
   {
-    length[1]= snprintf(str_data, sizeof(str_data), "MySQL%d", int_data);
+    length[1]= sprintf(str_data, "MySQL%d", int_data);
```
Comment (grooverdan) [inline tests/mysql_client_test.c:1491]: Why are you changing these `snprintf` back? Please don't.

File: tests/mysql_client_test.c [hunk 4]
```diff
@@ -13783,7 +13779,7 @@ static void test_bug8722()
   myquery(rc);
   /* Note: if you uncomment following block everything works fine */
 /*
-  rc= mysql_query(mysql, "sellect * from v1");
+  rc= mysql_query(mysql, "select * from v1");
```
Comment (grooverdan) [inline tests/mysql_client_test.c:13782]: obvious bug. But lets put that in a separate commit, maybe even just a separate PR. And is this test_bug8722 actually called?

File: tests/mysql_client_test.c [hunk 5]
```diff
@@ -17442,7 +17438,7 @@ static void test_bug30472()
   DIE_UNLESS(strcmp(character_set_name_4, "utf8mb3") == 0);
   DIE_UNLESS(strcmp(character_set_client_4, "utf8mb3") == 0);
   DIE_UNLESS(strcmp(character_set_results_4, "utf8mb3") == 0);
-  DIE_UNLESS(strcmp(collation_connnection_4, "utf8mb3_general_ci") == 0);
+  DIE_UNLESS(strcmp(collation_connnection_4, "utf8mb3_uca1400_ai_ci") == 0);
```
Comment (grooverdan) [inline tests/mysql_client_test.c:17441]: this is 11.x+ - so lets not change this. Its not relevant to your PR.

File: tests/mysql_client_test.c [hunk 6]
```diff
@@ -19441,15 +19437,15 @@ static void test_progress_reporting()
...
+  rc= mysql_query(conn, "alter table t1 add f1 int primary key auto_increment, lock=shared, order by f2");
   myquery(rc);
   if (!opt_silent)
     printf("Got progress_count: %u  stage: %u  max_stage: %u\n",
            progress_count, progress_stage, progress_max_stage);
   DIE_UNLESS(progress_count > 0 && progress_stage >=2 && progress_max_stage == 3);
 
   progress_stage= progress_max_stage= progress_count= 0;
-  rc= mysql_query(conn, "create index f2 on t1 (f2)");
+  rc= mysql_query(conn, "create index f2 on t1 (f2) lock=shared");
```
Comment (grooverdan) [inline tests/mysql_client_test.c:19448]: not relevant to the fix of MDEV-39466. Omit and process separately if required.

File: tests/mysql_client_test.c [hunk 7]
```diff
@@ -19462,7 +19458,7 @@ static void test_progress_reporting()
   if (!opt_silent)
     printf("Got progress_count: %u  stage: %u  max_stage: %u\n",
            progress_count, progress_stage, progress_max_stage);
-  DIE_UNLESS(progress_count > 0 && progress_stage >=2 && progress_max_stage == 2);
+  DIE_UNLESS(progress_count > 0 && progress_stage >=2 && progress_max_stage == 4);
```
Comment (grooverdan) [inline tests/mysql_client_test.c:19461]: separate PR please. Optionally take the spelling fixes there too.

File: tests/mysql_client_test.c [hunk 8]
```diff
@@ -23562,9 +23700,12 @@ static struct my_tests_st my_tests[]= {
...
+#ifndef EMBEDDED_LIBRARY
+  { "test_mdev_36080", test_mdev_36080},
   { "test_mdev35953", test_mdev35953 },
+#endif
   { 0, 0 }
 };
 
 
-static struct my_tests_st *get_my_tests() { return my_tests; }
+static struct my_tests_st *get_my_tests() { return my_tests; }
```
Comment (grooverdan) [inline tests/mysql_client_test.c:23711]: Ensure your editor doesn't remove this.

**Review by grooverdan — CHANGES_REQUESTED** [review]
Can you rebase the last commit of yours back to the origin/10.11 branch to include just your changes relevant to the off-by-one overflow and `git push --force` after amending the review comments into a single commit. Please add the MDEV-39466 to the header of your commit.

**Comment by grooverdan** [comment]
Sorry - it seems this fix was already in progress on a different bug report based on our earlier draft.

Feel free to take remaining fixing to a new PR.

https://jira.mariadb.org/browse/MDEV-39466
